modules/systemdstates/collect.go

```diff
@@ -0,0 +1,89 @@
+package systemdstates
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+)
+
+const (
+	inactive = 0
+	active   = 1
+	failed   = 2
+)
+
+func (s *SystemdStates) collect() (map[string]int64, error) {
+
+	var err error
+	conn, err := dbus.New()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	allUnits, err := conn.ListUnits()
+	if err != nil {
+		return nil, err
+	}
+
+	mx := make(map[string]int64)
+
+	units := s.filterUnits(allUnits)
+	for _, unit := range units {
+
+		ut, err := s.unitType(unit.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		chartID := fmt.Sprintf("systemd_%s_active_state", ut)
+		chart := s.charts.Get(chartID)
+		if !chart.HasDim(unit.Name) {
+			_ = chart.AddDim(&Dim{ID: unit.Name})
+		}
+
+		state := -1
+		if unit.ActiveState == "active" {
+			state = active
+		}
+		if unit.ActiveState == "inactive" {
+			state = inactive
+		}
+		if unit.ActiveState == "failed" {
+			state = failed
+		}
+		mx[unit.Name] = int64(state)
+	}
+
+	return mx, nil
+}
+
+func (s SystemdStates) filterUnits(units []dbus.UnitStatus) []dbus.UnitStatus {
+
+	filtered := make([]dbus.UnitStatus, 0, len(units))
+	for _, unit := range units {
+
+		if s.unitsMatcher.MatchString(unit.Name) && unit.LoadState == "loaded" {
+			filtered = append(filtered, unit)
+		}
+	}
+
+	return filtered
+}
```

Comment:
i suggest to use this approach for filtering, we can use same slice

https://github.com/netdata/go.d.plugin/blob/b1e849a145abe8e620e49fb9b884ce70ef564d6e/modules/elasticsearch/collect.go#L286-L296

> s.unitsMatcher.MatchString(unit.Name) && unit.LoadState == "loaded"

i think `unit.LoadState == "loaded"` should be first